### PR TITLE
web/storagenode: egress chart implemented

### PR DESCRIPTION
**Pull request by VitaliiShpital**
What:
API for new charts extended
egress chart for exact satellite and for all satellites implemented

Why: SNO Dashboard enhancements

Please describe the tests:
 - Test 1:
 - Test 2:
 
Please describe the performance impact:

## Code Review Checklist (to be filled out by reviewer)
 - [x] NEW: Are there any Satellite database migrations? Are they forwards _and_ backwards compatible? 
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: web/storagenode/src/app/components/EgressChart.vue [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+@Component ({
+    components: {
+        VChart,
+    },
+})
+export default class EgressChart extends Vue {
+    private readonly TOOLTIP_OPACITY: string = '1';
+    private readonly TOOLTIP_POSITION: string = 'absolute';
+
+    private get allBandwidth(): EgressBandwidthUsed[] {
```
Comment (crawter) [inline web/storagenode/src/app/components/EgressChart.vue]: EgressBandwidthUsed should be BandwidthUsed i think

File: web/storagenode/src/app/components/EgressChart.vue [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+<template>
+    <div class="chart">
+        <p class="egress-bandwidth-chart__data-dimension">{{chartDataDimension}}</p>
+        <VChart
+            id="egress-bandwidth-chart"
+            :chart-data="chartData"
+            :width="400"
+            :height="240"
+            :tooltip-constructor="egressBandwidthTooltip"
```
Comment (crawter) [inline web/storagenode/src/app/components/EgressChart.vue]: egressBandwidthTooltip - egressTooltip?

File: web/storagenode/src/storagenode/api/storagenode.ts [hunk 3]
```diff
@@ -69,14 +78,27 @@ export class SNOApi {
             return new BandwidthUsed(egress, ingress, new Date(bandwidth.intervalStart));
         });
 
+        const egressBandwidthDaily: EgressBandwidthUsed[] =  bandwidthDailyJson.map((bandwidth: any) => {
```
Comment (crawter) [inline web/storagenode/src/storagenode/api/storagenode.ts]: egressDaily

File: web/storagenode/src/storagenode/api/storagenode.ts [hunk 4]
```diff
@@ -69,14 +78,27 @@ export class SNOApi {
...
+            const egress = new Egress(bandwidth.egress.audit, bandwidth.egress.repair, bandwidth.egress.usage);
+
+            return new EgressBandwidthUsed(egress, new Date(bandwidth.intervalStart));
+        });
+
+        const ingressBandwidthDaily: IngressBandwidthUsed[] =  bandwidthDailyJson.map((bandwidth: any) => {
```
Comment (crawter) [inline web/storagenode/src/storagenode/api/storagenode.ts]: ingressDaily

File: web/storagenode/src/storagenode/api/storagenode.ts [hunk 5]
```diff
@@ -100,6 +122,19 @@ export class SNOApi {
             return new BandwidthUsed(egress, ingress, new Date(bandwidth.intervalStart));
         });
 
-        return new Satellites(storageDaily, bandwidthDaily, json.storageSummary, json.bandwidthSummary);
+        const egressBandwidthDaily: EgressBandwidthUsed[] =  bandwidthDailyJson.map((bandwidth: any) => {
```
Comment (crawter) [inline web/storagenode/src/storagenode/api/storagenode.ts]: egressDaily

File: web/storagenode/src/storagenode/api/storagenode.ts [hunk 6]
```diff
@@ -100,6 +122,19 @@ export class SNOApi {
...
+            const egress = new Egress(bandwidth.egress.audit, bandwidth.egress.repair, bandwidth.egress.usage);
+
+            return new EgressBandwidthUsed(egress, new Date(bandwidth.intervalStart));
+        });
+
+        const ingressBandwidthDaily: IngressBandwidthUsed[] =  bandwidthDailyJson.map((bandwidth: any) => {
```
Comment (crawter) [inline web/storagenode/src/storagenode/api/storagenode.ts]: const ingressDaily: IngressUsed

File: web/storagenode/src/storagenode/api/storagenode.ts [hunk 7]
```diff
@@ -100,6 +122,19 @@ export class SNOApi {
...
+            const ingress = new Ingress(bandwidth.ingress.repair, bandwidth.ingress.usage);
+
+            return new IngressBandwidthUsed(ingress, new Date(bandwidth.intervalStart));
```
Comment (crawter) [inline web/storagenode/src/storagenode/api/storagenode.ts]: IngressUsed

File: web/storagenode/src/storagenode/satellite.ts [hunk 8]
```diff
@@ -9,8 +9,12 @@ export class Satellite {
         public id: string,
         public storageDaily: Stamp[],
         public bandwidthDaily: BandwidthUsed[],
+        public egressBandwidthDaily: EgressBandwidthUsed[],
```
Comment (crawter) [inline web/storagenode/src/storagenode/satellite.ts]: same here

File: web/storagenode/src/storagenode/satellite.ts [hunk 9]
```diff
@@ -102,13 +106,75 @@ export class BandwidthUsed {
...
+        now.setDate(date);
+
+        return new EgressBandwidthUsed(new Egress(0, 0, 0), now);
+    }
+}
+
+/**
+ * IngressBandwidthUsed stores ingress bandwidth usage information over the period of time
+ */
+export class IngressBandwidthUsed {
```
Comment (crawter) [inline web/storagenode/src/storagenode/satellite.ts]: IngressUsed

File: storagenode/bandwidth/db_test.go [hunk 10]
```diff
@@ -40,6 +40,28 @@ var (
 	}
 )
 
+var (
+	egressActions = []pb.PieceAction{
+		pb.PieceAction_GET,
+		pb.PieceAction_GET_AUDIT,
+		pb.PieceAction_GET_REPAIR,
+
+		pb.PieceAction_GET,
```
Comment (crawter) [inline storagenode/bandwidth/db_test.go]: can't understand why do we need 2 copies of this values

File: web/storagenode/src/app/components/BandwidthChart.vue [hunk 11]
```diff
@@ -137,8 +147,13 @@ export default class BandwidthChart extends Vue {
...
             tooltipEl.style.opacity = this.TOOLTIP_OPACITY;
             tooltipEl.style.position = this.TOOLTIP_POSITION;
-            tooltipEl.style.left = position.left + tooltipModel.caretX + 'px';
-            tooltipEl.style.top = position.top + window.pageYOffset + tooltipModel.caretY + 'px';
+            tooltipEl.style.left = position.left + tooltipModel.caretX - 125 + 'px';
+            tooltipEl.style.bottom = position.bottom + window.pageYOffset - tooltipModel.caretY - 83 + 'px';
+
+            tooltipArrow.style.opacity = this.TOOLTIP_OPACITY;
+            tooltipArrow.style.position = this.TOOLTIP_POSITION;
+            tooltipArrow.style.left = position.left + tooltipModel.caretX - 24 + 'px';
```
Comment (NikolaiYurchenko) [inline web/storagenode/src/app/components/BandwidthChart.vue]: `${position.left + tooltipModel.caretX - 24}px`

File: web/storagenode/src/app/components/EgressChart.vue [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+    private get allBandwidth(): EgressUsed[] {
+        return ChartUtils.populateEmptyBandwidth(this.$store.state.node.egressChartData);
+    }
+
+    public get chartDataDimension(): string {
+        if (!this.$store.state.node.egressChartData.length) {
+            return 'Bytes';
+        }
+
+        return ChartUtils.getChartDataDimension(this.allBandwidth.map((elem) => {
```
Comment (NikolaiYurchenko) [inline web/storagenode/src/app/components/EgressChart.vue]: can be simplified elem => elem.summary()

File: storagenode/bandwidth/usage.go [hunk 13]
```diff
@@ -16,12 +16,21 @@ import (
...
-	// MonthSummary returns summary of the current months bandwidth usages
+	// MonthSummary returns summary of the current months bandwidth usages.
 	MonthSummary(ctx context.Context) (int64, error)
 	Rollup(ctx context.Context) (err error)
+	// Summary returns summary of bandwidth usages.
 	Summary(ctx context.Context, from, to time.Time) (*Usage, error)
+	// EgressSummary returns summary of egress bandwidth usages.
+	EgressSummary(ctx context.Context, from, to time.Time) (*Usage, error)
+	// IngressSummary returns summary of ingress bandwidth usages.
+	IngressSummary(ctx context.Context, from, to time.Time) (*Usage, error)
```
Comment (egonelbre) [inline storagenode/bandwidth/usage.go]: I wonder whether these would've been clearer and less requests to the database by having:

```
type Usages struct {
    Bandwidth Usage
    Egress Usage
    Ingress Usage
}

Summary(ctx context.Context, from, to time.Time) (*Usages, error)
```

**Pull request merged**